Set containerd pdeathsig

cpuguy83 opened the pull request:
Currently when starting containerd, docker just sets containerd's stdio
to docker's stdio.
If the docker daemon is restarted there is no way to re-attach and we
lose containerd's logs.

This change sets up fifo's for containerd's stdio so docker can
re-attach to containerd's stdio and not lose logs.

This doesn't seem like a perfect solution.
One caveat of this is that logrus (which containerd uses) will not use pretty mode for the log output since it will never be a terminal out with this method.

cpuguy83 commented:
ping @tonistiigi

tonistiigi commented:
On daemon shutdown, containerd is shut down also(if it was started by docker). The daemon crash case seems to be unhandled atm, so maybe we should set parent death signal instead. I'm actually quite surprised that we don't set it. @mlaventure did we remove it for some reason?

This solution will block containerd if daemon is not reading on the fifos, after the kernel buffer fills up.

mlaventure commented:
@tonistiigi 

That was an oversight from my part. I've never put the code to have `containerd` die if the daemon crash. On a clean `docker` shutdown, docker will take care of shutting down `containerd` though, preventing the issue described.

I agree with tonis that setting `cmd.SysProcAttr.Pdeathsig` to  `KILL` when starting `containerd` ourselves would probably be a "cleaner" solution.

cpuguy83 committed:
Set Pdeathsig for containerd on SIGKILL

Makes sure containerd exits (when started by docker) if docker gets
SIGKILL'd.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
I've updated this to set pdeathsig to `syscall.SIGKILL`

mlaventure commented:
LGTM, IANAM

tonistiigi commented:
LGTM

thaJeztah commented:
LGTM
